pageserver/src/tenant/timeline.rs

```diff
@@ -2663,6 +2674,30 @@ impl Timeline {
         let layers = self.layers.read().unwrap();
 
         let mut max_deltas = 0;
+        let wanted_image_layers = self.wanted_image_layers.lock().unwrap();
+        if let Some((cutoff_lsn, wanted)) = &*wanted_image_layers {
+            let img_range =
+                partition.ranges.first().unwrap().start..partition.ranges.last().unwrap().end;
+            if wanted.overlaps(&img_range) {
+                //
+                // gc_timeline only pays attention to image layers that are older than the GC cutoff,
+                // but create_image_layers creates image layers at last-record-lsn.
+                // So it's possible that gc_timeline decides that it wants new image layer to be created for a key range,
+                // and on next compaction create_image_layers creates the image layer.
+                // But on next GC cycle, gc_timeline still wants the new image layer to be created,
+                // because the newly created image layer doesn't help to remove the delta layer,
+                // until the newly created image layer falls off the PITR horizon.
+                //
+                // So we should check if image layer beyond cutoff LSN already exists.
+                if !layers.image_layer_exists(&img_range, &(Lsn::min(lsn, *cutoff_lsn)..lsn + 1))? {
+                    debug!(
+                        "Force generation of layer {}-{} wanted by GC, cutoff={}, lsn={})",
+                        img_range.start, img_range.end, cutoff_lsn, lsn
+                    );
+                    return Ok(true);
+                }
+            }
+        }
```

Comment:
Let's wrap this in a block `{ ... }`, so that the lock is dropped as soon as we have made the check.

Reply:
Done